### PR TITLE
Custom convert incoming long for date and datetime types

### LINKED ISSUE #35946: Failure in date time conversion during mssql first sync: Long cannot be cast to class java.sql.Timestamp
### Connector Name

source-mssql

### Connector Version

4.0.0

### What step the error happened?

None

### Relevant information

Discovered in [workspace](https://cloud.airbyte.com/workspaces/4813479c-9dd8-41a9-8ecf-5080c370ae4f/connections/060d7920-6931-4f48-9163-da5cf72a7c40/job-history#9073636::4)

Error while reading for the first time 

### Relevant log output

```shell
2024-03-06 18:05:59 source > ERROR i.d.r.RelationalSnapshotChangeEventSource(doExecute):179 Error during snapshot io.debezium.DebeziumException: java.lang.ClassCastException: class java.lang.Long cannot be cast to class java.sql.Timestamp (java.lang.Long is in module java.base of loader 'bootstrap'; java.sql.Timestamp is in module java.sql of loader 'platform')
        at io.debezium.relational.RelationalSnapshotChangeEventSource.lambda$createSchemaChangeEventsForTables$3(RelationalSnapshotChangeEventSource.java:387) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        at io.debezium.pipeline.EventDispatcher.dispatchSchemaChangeEvent(EventDispatcher.java:367) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        at io.debezium.relational.RelationalSnapshotChangeEventSource.createSchemaChangeEventsForTables(RelationalSnapshotChangeEventSource.java:382) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        at io.debezium.relational.RelationalSnapshotChangeEventSource.doExecute(RelationalSnapshotChangeEventSource.java:154) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        at io.debezium.pipeline.source.AbstractSnapshotChangeEventSource.execute(AbstractSnapshotChangeEventSource.java:92) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        at io.debezium.pipeline.ChangeEventSourceCoordinator.doSnapshot(ChangeEventSourceCoordinator.java:253) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        at io.debezium.pipeline.ChangeEventSourceCoordinator.doSnapshot(ChangeEventSourceCoordinator.java:237) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        at io.debezium.connector.sqlserver.SqlServerChangeEventSourceCoordinator.executeChangeEventSources(SqlServerChangeEventSourceCoordinator.java:82) ~[debezium-connector-sqlserver-2.4.0.Final.jar:2.4.0.Final]
        at io.debezium.pipeline.ChangeEventSourceCoordinator.lambda$start$0(ChangeEventSourceCoordinator.java:137) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) ~[?:?]
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: java.lang.ClassCastException: class java.lang.Long cannot be cast to class java.sql.Timestamp (java.lang.Long is in module java.base of loader 'bootstrap'; java.sql.Timestamp is in module java.sql of loader 'platform')
        at io.airbyte.integrations.source.mssql.MssqlDebeziumConverter.lambda$registerDatetime$3(MssqlDebeziumConverter.java:133) ~[io.airbyte.airbyte-integrations.connectors-source-mssql.jar:?]
        at io.debezium.relational.CustomConverterRegistry.lambda$getValueConverter$1(CustomConverterRegistry.java:147) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        at io.debezium.relational.TableSchemaBuilder.addField(TableSchemaBuilder.java:406) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        at io.debezium.relational.TableSchemaBuilder.lambda$create$2(TableSchemaBuilder.java:149) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184) ~[?:?]
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179) ~[?:?]
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708) ~[?:?]
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[?:?]
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151) ~[?:?]
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174) ~[?:?]
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596) ~[?:?]
        at io.debezium.relational.TableSchemaBuilder.create(TableSchemaBuilder.java:147) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        at io.debezium.relational.RelationalDatabaseSchema.buildAndRegisterSchema(RelationalDatabaseSchema.java:122) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        at io.debezium.connector.sqlserver.SqlServerDatabaseSchema.applySchemaChange(SqlServerDatabaseSchema.java:53) ~[debezium-connector-sqlserver-2.4.0.Final.jar:2.4.0.Final]
        at io.debezium.pipeline.EventDispatcher$SchemaChangeEventReceiver.schemaChangeEvent(EventDispatcher.java:622) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        at io.debezium.relational.RelationalSnapshotChangeEventSource.lambda$createSchemaChangeEventsForTables$3(RelationalSnapshotChangeEventSource.java:384) ~[debezium-core-2.4.0.Final.jar:2.4.0.Final]
        ... 13 more
```

### Contribute

- [ ] Yes, I want to contribute

### PR DESCRIPTION
**Pull request by rodireich**
Date types included in tables as default values are discovered in debezium with an input of either epoch milliseconds (for Datetime types) or epoch days (for date type).
Our custom converter couldn't convert this input.
Added correct conversion for schema history phase.

Actual record Dates are still seen as an incoming TimeStamp.

### CODE REVIEW

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlDebeziumConverter.java [hunk 1]
```diff
@@ -129,9 +125,21 @@ private void registerDatetime(final RelationalColumn field,
           if (Objects.isNull(input)) {
             return DebeziumConverterUtils.convertDefaultValue(field);
           }
+          if (input instanceof final Timestamp d) {
+            final LocalDateTime localDateTime = d.toLocalDateTime();
+            return localDateTime.format(DateTimeFormatter.ofPattern(DATETIME_FORMAT_MICROSECONDS));
+          }
+
+          if (input instanceof final Long d) {
```
Comment (evantahler) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlDebeziumConverter.java:133]: Is it the case that for Mssql that all timestamp 'numbers' are _always_ with nanosecond precision?

Comment (rodireich) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlDebeziumConverter.java:133]: We're only seeing this Long input when values are converted as part of acquiring schema history.
Dates on actual row records are still seen as incoming `TimeStamp`. None of this is documented anywhere as far as I can tell.

We need to take a close look at some point as this area is uneven in terms of implementation.
And [some of the code](https://github.com/airbytehq/airbyte/commit/56db8065e9fbfbf054303c9d1a8c9694547dd02a) here is almost three years old.

**Review by akashkulk — APPROVED** [review]

**Pull request merged**